File: src/commonly_cited/parsing.py
"""Reference-list segmentation and lightweight feature extraction."""

from __future__ import annotations

import re
import unicodedata
from typing import TYPE_CHECKING

from .models import Reference
from .normalization import normalize_doi, normalize_surname, unique_preserving_order

if TYPE_CHECKING:
    from collections.abc import Iterable

_DOI_RE = re.compile(r"\b10\.\d{4,9}/[-._;()/:A-Z0-9]+", re.IGNORECASE)
_YEAR_RE = re.compile(r"(?<!\d)(?:18|19|20|21)\d{2}[a-z]?(?!\d)", re.IGNORECASE)
_ET_AL_RE = re.compile(r"\bet\s+al\.?\b", re.IGNORECASE)
_NUMBERED_START_RE = re.compile(
    r"^\s*(?:\[(?P<bracket>\d{1,3})\]|\((?P<paren>\d{1,3})\)|"
    r"(?P<plain>\d{1,3})[.)])\s*(?P<body>.+?)\s*$"
)
_INLINE_NUMBER_RE = re.compile(
    r"(?<!\w)(?:\[(?P<bracket>\d{1,3})\]|\((?P<paren>\d{1,3})\)|"
    r"(?P<plain>\d{1,3})[.)])\s+"
)
_DASH_CHARACTERS = "-\u2010\u2011\u2012\u2013\u2014\u2015"
_BIBTEX_START_RE = re.compile(
    r"^\s*@(?:article|book|inproceedings|misc|phdthesis|"
    r"mastersthesis|techreport|incollection|proceedings)\s*\{",
    re.IGNORECASE,
)
_RIS_START_RE = re.compile(r"^TY\s{0,2}-\s*", re.IGNORECASE)
_RIS_END_RE = re.compile(r"^ER\s{0,2}-\s*", re.IGNORECASE)
_HEADING_RE = re.compile(r"^\s*(?:references|bibliography|works\s+cited)\s*:?[\s]*$", re.IGNORECASE)
_BULLET_RE = re.compile(r"^\s*(?:[-*•‣▪◦])\s+")

_SURNAME_COMMA_RE = re.compile(
    r"(?<![\w'-])([A-ZÀ-ÖØ-Þ][A-Za-zÀ-ÖØ-öø-ÿ'\u2019\-]{1,})\s*,\s*"
    r"(?:[A-ZÀ-ÖØ-Þ](?:\.|[a-zà-öø-ÿ]+))"
)
_INITIALS_SURNAME_RE = re.compile(
    r"(?:\b[A-ZÀ-ÖØ-Þ]\.?\s*){1,4}"
    r"([A-ZÀ-ÖØ-Þ][A-Za-zÀ-ÖØ-öø-ÿ'\u2019\-]{1,})\b"
)
_SURNAME_INITIALS_RE = re.compile(
    r"\b([A-ZÀ-ÖØ-Þ][A-Za-zÀ-ÖØ-öø-ÿ'\u2019\-]{1,})\s+"
    r"(?:[A-ZÀ-ÖØ-Þ]\.?\s*){1,4}(?=,|;|\s+(?:and|&)\b|$)"
)
_ET_AL_SURNAME_RE = re.compile(
    r"\b([A-ZÀ-ÖØ-Þ][A-Za-zÀ-ÖØ-öø-ÿ'\u2019\-]{1,})\s+(?:,\s*)?et\s+al\.?",
    re.IGNORECASE,
)


def _clean_text(text: str) -> str:
    text = unicodedata.normalize("NFKC", text).replace("\u00ad", "")
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    # Repair words broken by a PDF line-wrap while leaving ordinary hyphens intact.
    return re.sub(r"(?<=[A-Za-z])-[ \t]*\n[ \t]*(?=[a-z])", "", text)


def _clean_reference(value: str) -> str:
    value = _BULLET_RE.sub("", value.strip())
    value = re.sub(r"\s+", " ", value)
    return value.strip()


def extract_doi(value: str) -> str | None:
    """Extract the first DOI from noisy reference text."""
    match = _DOI_RE.search(value)
    return normalize_doi(match.group(0)) if match else None


def extract_years(value: str) -> tuple[int, ...]:
    """Extract plausible publication years from a reference."""
    years: list[int] = []
    for match in _YEAR_RE.finditer(value):
        year = int(match.group(0)[:4])
        if year not in years:
            years.append(year)
    return tuple(years)


def extract_visible_surnames(value: str) -> tuple[str, ...]:
    """Extract visible author surnames for candidate scoring.

    This is intentionally conservative. The metadata provider, rather than this
    heuristic, supplies the final author list.
    """
    year_match = _YEAR_RE.search(value)
    prefix = value[: year_match.start()] if year_match else value[:300]
    prefix = _NUMBERED_START_RE.sub(lambda match: match.group("body"), prefix)
    prefix = prefix.split(".", maxsplit=1)[0] if len(prefix) > 180 else prefix

    candidates: list[str] = []
    candidates.extend(match.group(1) for match in _SURNAME_COMMA_RE.finditer(prefix))
    candidates.extend(match.group(1) for match in _INITIALS_SURNAME_RE.finditer(prefix))
    candidates.extend(match.group(1) for match in _SURNAME_INITIALS_RE.finditer(prefix))
    candidates.extend(match.group(1) for match in _ET_AL_SURNAME_RE.finditer(prefix))

    normalized = (normalize_surname(candidate) for candidate in candidates)
    return unique_preserving_order(normalized)


def _split_bibtex(text: str) -> list[str]:
    entries: list[str] = []
    current: list[str] = []
    depth = 0
    in_entry = False
    for line in text.splitlines():
        if not in_entry and _BIBTEX_START_RE.match(line):
            in_entry = True
            current = []
            depth = 0
        if in_entry:
            current.append(line)
            depth += line.count("{") - line.count("}")
            if depth <= 0 and current:
                entries.append(_clean_reference("\n".join(current)))
                current = []
                in_entry = False
    if current:
        entries.append(_clean_reference("\n".join(current)))
    return entries


def _split_ris(text: str) -> list[str]:
    entries: list[str] = []
    current: list[str] = []
    for line in text.splitlines():
        if _RIS_START_RE.match(line):
            if current:
                entries.append(_clean_reference("\n".join(current)))
            current = [line]
        elif current:
            current.append(line)
            if _RIS_END_RE.match(line):
                entries.append(_clean_reference("\n".join(current)))
                current = []
    if current:
        entries.append(_clean_reference("\n".join(current)))
    return entries


def _split_inline_numbered(text: str) -> list[str]:
    matches = [
        match
        for match in _INLINE_NUMBER_RE.finditer(text)
        if not (
            match.group("plain")
            and text[: match.start()].rstrip().endswith(tuple(_DASH_CHARACTERS))
        )
    ]
    if len(matches) < 2:
        return []
    # Inline markers are only credible when the first one starts the list. This
    # avoids mistaking journal volumes or page numbers such as ``1.`` and ``2.``
    # at line endings for bibliography numbering.
    prefix = text[: matches[0].start()].strip()
    if prefix and not _HEADING_RE.match(prefix):
        return []

    first = matches[0]
    style = next(name for name in ("bracket", "paren", "plain") if first.group(name))
    expected_number = int(first.group(style)) + 1
    sequence = [first]
    for match in matches[1:]:
        value = match.group(style)
        if value is not None and int(value) == expected_number:
            sequence.append(match)
            expected_number += 1
    matches = sequence
    if len(matches) < 2:
        return []

    entries: list[str] = []
    for index, match in enumerate(matches):
        start = match.end()
        end = matches[index + 1].start() if index + 1 < len(matches) else len(text)
        entry = _clean_reference(text[start:end])
        if entry:
            entries.append(entry)
    return entries


def _split_numbered_lines(lines: list[str]) -> list[str]:
    entries: list[str] = []
    current: list[str] = []
    found_marker = False
    for line in lines:
        match = _NUMBERED_START_RE.match(line)
        if match:
            found_marker = True
            if current:
                entries.append(_clean_reference(" ".join(current)))
            current = [match.group("body")]
        elif current and line.strip():
            current.append(line.strip())
    if current:
        entries.append(_clean_reference(" ".join(current)))
    return entries if found_marker else []


def _looks_like_reference(value: str) -> bool:
    cleaned = _clean_reference(value)
    return len(cleaned) >= 25 and (extract_doi(cleaned) is not None or bool(extract_years(cleaned)))


def _looks_like_reference_start(value: str) -> bool:
    cleaned = _clean_reference(value)
    if len(cleaned) < 12:
        return False
    first_year = _YEAR_RE.search(cleaned[:180])
    if first_year is None:
        return False
    prefix = cleaned[: first_year.start()]
    return bool("," in prefix or _ET_AL_RE.search(prefix) or _INITIALS_SURNAME_RE.search(prefix))


def _split_blank_blocks(lines: list[str]) -> list[str]:
    blocks: list[str] = []
    current: list[str] = []
    for line in lines:
        if line.strip():
            current.append(line.strip())
        elif current:
            blocks.append(_clean_reference(" ".join(current)))
            current = []
    if current:
        blocks.append(_clean_reference(" ".join(current)))
    plausible = sum(_looks_like_reference(block) for block in blocks)
    if len(blocks) >= 2 and plausible >= len(blocks) / 2:
        return blocks
    return []


def _split_semicolon_candidates(text: str) -> list[str]:
    parts = [_clean_reference(part) for part in re.split(r"\s*;\s*(?=[A-Z\[])", text)]
    parts = [part for part in parts if part]
    if len(parts) >= 2 and all(_looks_like_reference(part) for part in parts):
        return parts
    return []


def split_references(text: str) -> list[str]:
    """Split a poorly formatted bibliography into individual references."""
    text = _clean_text(text)
    lines = [line for line in text.splitlines() if not _HEADING_RE.match(line)]
    text_without_heading = "\n".join(lines).strip()
    if not text_without_heading:
        return []

    if _BIBTEX_START_RE.search(text_without_heading):
        entries = _split_bibtex(text_without_heading)
        if entries:
            return entries

    if any(_RIS_START_RE.match(line) for line in lines):
        entries = _split_ris(text_without_heading)
        if entries:
            return entries

    entries = _split_inline_numbered(text_without_heading)
    if entries:
        return entries

    entries = _split_numbered_lines(lines)
    if entries:
        return entries

    entries = _split_blank_blocks(lines)
    if entries:
        return entries

    nonempty = [_clean_reference(line) for line in lines if line.strip()]
    plausible_lines = sum(_looks_like_reference(line) for line in nonempty)
    if len(nonempty) >= 2 and plausible_lines >= 0.7 * len(nonempty):
        return nonempty

    entries = []
    current: list[str] = []
    for line in nonempty:
        current_is_reference = _looks_like_reference(" ".join(current))
        if current and _looks_like_reference_start(line) and current_is_reference:
            entries.append(_clean_reference(" ".join(current)))
            current = [line]
        else:
            current.append(line)
    if current:
        entries.append(_clean_reference(" ".join(current)))
    if len(entries) >= 2:
        return entries

    semicolon_entries = _split_semicolon_candidates(text_without_heading)
    if semicolon_entries:
        return semicolon_entries

    return [_clean_reference(text_without_heading)]


def parse_references(text: str) -> list[Reference]:
    """Parse bibliography text into references with matching features."""
    references: list[Reference] = []
    for index, raw in enumerate(split_references(text), start=1):
        references.append(
            Reference(
                index=index,
                raw=raw,
                doi=extract_doi(raw),
                years=extract_years(raw),
                visible_surnames=extract_visible_surnames(raw),
                has_et_al=bool(_ET_AL_RE.search(raw)),
            )
        )
    return references


def join_reference_lines(lines: Iterable[str]) -> str:
    """Public helper for normalizing a stream of reference lines."""
    return _clean_reference(" ".join(line.strip() for line in lines if line.strip()))

File: web/src/lib/parsing.ts
import {
  normalizeDoi,
  normalizeSurname,
  uniquePreservingOrder,
} from "./normalization";
import type { Reference } from "./types";

const DOI_RE = /\b10\.\d{4,9}\/[-._;()/:A-Z0-9]+/i;
const YEAR_RE = /(?<!\d)(?:18|19|20|21)\d{2}[a-z]?(?!\d)/gi;
const ET_AL_RE = /\bet\s+al\.?\b/i;
const NUMBERED_START_RE =
  /^\s*(?:\[(\d{1,3})\]|\((\d{1,3})\)|(\d{1,3})[.)])\s*(.+?)\s*$/;
const INLINE_NUMBER_RE =
  /(?<!\w)(?:\[(\d{1,3})\]|\((\d{1,3})\)|(\d{1,3})[.)])\s+/g;
const DASH_CHARACTERS = "-‐‑‒–—―";
const BIBTEX_START_RE =
  /^\s*@(?:article|book|inproceedings|misc|phdthesis|mastersthesis|techreport|incollection|proceedings)\s*\{/i;
const RIS_START_RE = /^TY\s{0,2}-\s*/i;
const RIS_END_RE = /^ER\s{0,2}-\s*/i;
const HEADING_RE = /^\s*(?:references|bibliography|works\s+cited)\s*:?\s*$/i;
const BULLET_RE = /^\s*(?:[-*•‣▪◦])\s+/;

const SURNAME_COMMA_RE =
  /(?<![\p{Letter}\p{Number}_'-])(\p{Lu}[\p{L}'’-]{1,})\s*,\s*(?:\p{Lu}(?:\.|\p{Ll}+))/gu;
const INITIALS_SURNAME_RE =
  /(?:\b\p{Lu}\.?\s*){1,4}(\p{Lu}[\p{L}'’-]{1,})\b/gu;
const SURNAME_INITIALS_RE =
  /\b(\p{Lu}[\p{L}'’-]{1,})\s+(?:\p{Lu}\.?\s*){1,4}(?=,|;|\s+(?:and|&)\b|$)/gu;
const ET_AL_SURNAME_RE =
  /\b(\p{Lu}[\p{L}'’-]{1,})\s+(?:,\s*)?et\s+al\.?/giu;

function cleanText(text: string): string {
  return text
    .normalize("NFKC")
    .replace(/\u00ad/g, "")
    .replace(/\r\n?/g, "\n")
    .replace(/(?<=[A-Za-z])-[ \t]*\n[ \t]*(?=[a-z])/g, "");
}

function cleanReference(value: string): string {
  return value.trim().replace(BULLET_RE, "").replace(/\s+/g, " ").trim();
}

export function extractDoi(value: string): string | null {
  const match = DOI_RE.exec(value);
  return match ? normalizeDoi(match[0]) : null;
}

export function extractYears(value: string): number[] {
  const years: number[] = [];
  for (const match of value.matchAll(YEAR_RE)) {
    const year = Number(match[0].slice(0, 4));
    if (!years.includes(year)) years.push(year);
  }
  return years;
}

function captures(value: string, pattern: RegExp): string[] {
  return [...value.matchAll(pattern)]
    .map((match) => match[1])
    .filter((item): item is string => item !== undefined);
}

export function extractVisibleSurnames(value: string): string[] {
  const yearMatch = new RegExp(YEAR_RE.source, "i").exec(value);
  let prefix = yearMatch ? value.slice(0, yearMatch.index) : value.slice(0, 300);
  const numbered = NUMBERED_START_RE.exec(prefix);
  if (numbered?.[4]) prefix = numbered[4];
  if (prefix.length > 180) prefix = prefix.split(".", 1)[0] ?? prefix;

  const candidates = [
    ...captures(prefix, SURNAME_COMMA_RE),
    ...captures(prefix, INITIALS_SURNAME_RE),
    ...captures(prefix, SURNAME_INITIALS_RE),
    ...captures(prefix, ET_AL_SURNAME_RE),
  ];
  return uniquePreservingOrder(candidates.map(normalizeSurname));
}

function splitBibtex(text: string): string[] {
  const entries: string[] = [];
  let current: string[] = [];
  let depth = 0;
  let inEntry = false;
  for (const line of text.split("\n")) {
    if (!inEntry && BIBTEX_START_RE.test(line)) {
      inEntry = true;
      current = [];
      depth = 0;
    }
    if (!inEntry) continue;
    current.push(line);
    depth += (line.match(/\{/g) ?? []).length - (line.match(/}/g) ?? []).length;
    if (depth <= 0 && current.length > 0) {
      entries.push(cleanReference(current.join("\n")));
      current = [];
      inEntry = false;
    }
  }
  if (current.length > 0) entries.push(cleanReference(current.join("\n")));
  return entries;
}

function splitRis(text: string): string[] {
  const entries: string[] = [];
  let current: string[] = [];
  for (const line of text.split("\n")) {
    if (RIS_START_RE.test(line)) {
      if (current.length > 0) entries.push(cleanReference(current.join("\n")));
      current = [line];
    } else if (current.length > 0) {
      current.push(line);
      if (RIS_END_RE.test(line)) {
        entries.push(cleanReference(current.join("\n")));
        current = [];
      }
    }
  }
  if (current.length > 0) entries.push(cleanReference(current.join("\n")));
  return entries;
}

function splitInlineNumbered(text: string): string[] {
  const matches = [...text.matchAll(INLINE_NUMBER_RE)].filter((match) => {
    if (match[3] === undefined || match.index === undefined) return true;
    const preceding = text.slice(0, match.index).trimEnd().at(-1);
    return preceding === undefined || !DASH_CHARACTERS.includes(preceding);
  });
  if (matches.length < 2) return [];
  const first = matches[0];
  if (!first || first.index === undefined) return [];
  const prefix = text.slice(0, first.index).trim();
  if (prefix && !HEADING_RE.test(prefix)) return [];

  const styleIndex = [1, 2, 3].find((index) => first[index] !== undefined);
  if (styleIndex === undefined) return [];
  let expectedNumber = Number(first[styleIndex]) + 1;
  const sequence = [first];
  for (const match of matches.slice(1)) {
    if (match[styleIndex] !== undefined && Number(match[styleIndex]) === expectedNumber) {
      sequence.push(match);
      expectedNumber += 1;
    }
  }
  if (sequence.length < 2) return [];

  const entries: string[] = [];
  for (let index = 0; index < sequence.length; index += 1) {
    const match = sequence[index];
    if (!match || match.index === undefined) continue;
    const next = sequence[index + 1];
    const entry = cleanReference(
      text.slice(match.index + match[0].length, next?.index ?? text.length),
    );
    if (entry) entries.push(entry);
  }
  return entries;
}

function splitNumberedLines(lines: string[]): string[] {
  const entries: string[] = [];
  let current: string[] = [];
  let foundMarker = false;
  for (const line of lines) {
    const match = NUMBERED_START_RE.exec(line);
    if (match) {
      foundMarker = true;
      if (current.length > 0) entries.push(cleanReference(current.join(" ")));
      current = match[4] ? [match[4]] : [];
    } else if (current.length > 0 && line.trim()) {
      current.push(line.trim());
    }
  }
  if (current.length > 0) entries.push(cleanReference(current.join(" ")));
  return foundMarker ? entries : [];
}

function looksLikeReference(value: string): boolean {
  const cleaned = cleanReference(value);
  return cleaned.length >= 25 && (extractDoi(cleaned) !== null || extractYears(cleaned).length > 0);
}

function looksLikeReferenceStart(value: string): boolean {
  const cleaned = cleanReference(value);
  if (cleaned.length < 12) return false;
  const match = new RegExp(YEAR_RE.source, "i").exec(cleaned.slice(0, 180));
  if (!match) return false;
  const prefix = cleaned.slice(0, match.index);
  return (
    prefix.includes(",") ||
    ET_AL_RE.test(prefix) ||
    new RegExp(INITIALS_SURNAME_RE.source, "u").test(prefix)
  );
}

function splitBlankBlocks(lines: string[]): string[] {
  const blocks: string[] = [];
  let current: string[] = [];
  for (const line of lines) {
    if (line.trim()) {
      current.push(line.trim());
    } else if (current.length > 0) {
      blocks.push(cleanReference(current.join(" ")));
      current = [];
    }
  }
  if (current.length > 0) blocks.push(cleanReference(current.join(" ")));
  const plausible = blocks.filter(looksLikeReference).length;
  return blocks.length >= 2 && plausible >= blocks.length / 2 ? blocks : [];
}

function splitSemicolonCandidates(text: string): string[] {
  const parts = text
    .split(/\s*;\s*(?=[A-Z]|\[)/)
    .map(cleanReference)
    .filter(Boolean);
  return parts.length >= 2 && parts.every(looksLikeReference) ? parts : [];
}

export function splitReferences(input: string): string[] {
  const text = cleanText(input);
  const lines = text.split("\n").filter((line) => !HEADING_RE.test(line));
  const textWithoutHeading = lines.join("\n").trim();
  if (!textWithoutHeading) return [];

  if (textWithoutHeading.split("\n").some((line) => BIBTEX_START_RE.test(line))) {
    const entries = splitBibtex(textWithoutHeading);
    if (entries.length > 0) return entries;
  }
  if (lines.some((line) => RIS_START_RE.test(line))) {
    const entries = splitRis(textWithoutHeading);
    if (entries.length > 0) return entries;
  }

  let entries = splitInlineNumbered(textWithoutHeading);
  if (entries.length > 0) return entries;
  entries = splitNumberedLines(lines);
  if (entries.length > 0) return entries;
  entries = splitBlankBlocks(lines);
  if (entries.length > 0) return entries;

  const nonempty = lines.map(cleanReference).filter(Boolean);
  const plausibleLines = nonempty.filter(looksLikeReference).length;
  if (nonempty.length >= 2 && plausibleLines >= 0.7 * nonempty.length) return nonempty;

  entries = [];
  let current: string[] = [];
  for (const line of nonempty) {
    if (
      current.length > 0 &&
      looksLikeReferenceStart(line) &&
      looksLikeReference(current.join(" "))
    ) {
      entries.push(cleanReference(current.join(" ")));
      current = [line];
    } else {
      current.push(line);
    }
  }
  if (current.length > 0) entries.push(cleanReference(current.join(" ")));
  if (entries.length >= 2) return entries;

  const semicolonEntries = splitSemicolonCandidates(textWithoutHeading);
  return semicolonEntries.length > 0
    ? semicolonEntries
    : [cleanReference(textWithoutHeading)];
}

export function parseReference(value: string, index = 1): Reference {
  const raw = cleanReference(value);
  return {
    index,
    raw,
    doi: extractDoi(raw),
    years: extractYears(raw),
    visibleSurnames: extractVisibleSurnames(raw),
    hasEtAl: ET_AL_RE.test(raw),
  };
}

export function parseReferences(text: string): Reference[] {
  return splitReferences(text).map((raw, offset) => parseReference(raw, offset + 1));
}

export function joinReferenceLines(lines: Iterable<string>): string {
  return cleanReference([...lines].filter((line) => line.trim()).join(" "));
}
